fvm/evm/emulator/emulator.go

```diff
@@ -233,6 +239,121 @@ func (proc *procedure) withdrawFrom(address types.Address, amount *big.Int) (*ty
 	return res, proc.commit()
 }
 
+// deployAt deploys a contract at the given target address
+// behaviour should be similar to what evm.create internal method does with
+// a few differences, don't need to check for previous forks given this
+// functionality was not available to anyone, we don't need to
+// follow snapshoting, given we do commit/revert style in this code base.
+// in the future we might optimize this method accepting deploy-ready byte codes
+// and skip interpreter call, gas calculations and many checks.
+func (proc *procedure) deployAt(
+	caller types.Address,
+	to types.Address,
+	data types.Code,
+	gasLimit uint64,
+	value *big.Int,
+) (*types.Result, error) {
+	res := &types.Result{
+		TxType: types.DirectCallTxType,
+	}
+	addr := to.ToCommon()
+
+	// precheck 1 - check balance of the source
+	if value.Sign() != 0 &&
+		!proc.evm.Context.CanTransfer(proc.state, caller.ToCommon(), value) {
+		return res, gethVM.ErrInsufficientBalance
+	}
+
+	// precheck 2 - ensure there's no existing contract is deployed at the address
+	contractHash := proc.state.GetCodeHash(addr)
+	if proc.state.GetNonce(addr) != 0 ||
+		(contractHash != (gethCommon.Hash{}) && contractHash != gethTypes.EmptyCodeHash) {
+		return res, gethVM.ErrContractAddressCollision
+	}
+
+	callerCommon := caller.ToCommon()
+	// setup caller if doesn't exist
+	if !proc.state.Exist(callerCommon) {
+		proc.state.CreateAccount(callerCommon)
+	}
+	// increment the nonce for the caller
+	proc.state.SetNonce(callerCommon, proc.state.GetNonce(callerCommon)+1)
+
+	// setup account
+	proc.state.CreateAccount(addr)
+	proc.state.SetNonce(addr, 1) // (EIP-158)
+	proc.evm.Context.Transfer(   // transfer value
```

Comment:
there is no protection against negative values being transfered, also I would avoid making a transfer if the value is 0, but in the first case I would return an error.

Reply:
I added some guards but the underlying layers handels negative or zero balance.